### PR TITLE
[stablehlo] Lower stablehlo.reduce_window cumsum to linalg_ext.scan

### PR DESCRIPTION
**Pull request by rsuderman**
Cumulative sums can more efficiently be represented as a `linalg_ext.scan` operation. Currently this only works for the forward cumulative sum however we should consider expanding it to a postfix sum in the future.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
> Cumulative sums can more efficiently be represented as a `linalg_ext.scan` operation. Currently this only works for the forward cumulative sum however we should consider expanding it to a postfix sum in the future.

I think that would need a separate operation. The size of the output is different... but maybe they can be managed in the same operation....

**Comment by rsuderman** [comment]
> > Cumulative sums can more efficiently be represented as a `linalg_ext.scan` operation. Currently this only works for the forward cumulative sum however we should consider expanding it to a postfix sum in the future.
> 
> I think that would need a separate operation. The size of the output is different... but maybe they can be managed in the same operation....

There is a chance I am mistaking about what is a prefix vs postfix sum. In this case I am just referring to a reversed prefix sum. E.g. `postfix(x) = reverse(prefixsum(reverse(x)))`

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalgExt.cpp [hunk 1]
```diff
@@ -439,6 +440,167 @@ struct ReverseOpConversion final
...
+        initDynDims.push_back(
+            rewriter.createOrFold<tensor::DimOp>(op.getLoc(), input0, i));
+      }
+    }
+
+    outputs.push_back(rewriter.create<tensor::EmptyOp>(
+        op.getLoc(), input0Ty.getShape(), input0Ty.getElementType(),
+        outputDynDims));
+
+    // TODO(suderman): Change this to something better
```
Comment (NatashaKnk) [inline compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalgExt.cpp]: This seems a bit vague

File: compiler/src/iree/compiler/InputConversion/StableHLO/test/stablehlo_to_linalg_ext.mlir [hunk 2]
```diff
@@ -614,3 +614,31 @@ func.func @chlo_top_k_float(%arg : tensor<16x16xf32>) -> (tensor<16x8xf32>, tens
 // CHECK:        %[[D7:.+]] = arith.cmpf ogt, %[[ARG1]], %[[ARG2]] : f32
 // CHECK:        iree_linalg_ext.yield %[[D7]] : i1
 // CHECK:        return %[[D6]]#0, %[[D6]]#1
+
```
Comment (NatashaKnk) [inline compiler/src/iree/compiler/InputConversion/StableHLO/test/stablehlo_to_linalg_ext.mlir:617]: nit: Maybe a bit too many new lines :)

**Review by NatashaKnk — APPROVED** [review]

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalgExt.cpp (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalgExt.cpp]: Yeah, I think this was a carry over as well. Deleted.

File: compiler/src/iree/compiler/InputConversion/StableHLO/test/stablehlo_to_linalg_ext.mlir (hunk 2 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/StableHLO/test/stablehlo_to_linalg_ext.mlir:617]: w   h   i   t   e   s   p   a   c   e

**Comment by pzread** [comment]
[build_test_all_macos_arm64](https://github.com/openxla/iree/actions/runs/6228083319/job/16904088683#logs) in postsubmit seems to be constantly failing on `iree/tests/e2e/stablehlo_ops/check_metal-spirv_metal_reduce_window.mlir` after this change:

https://github.com/openxla/iree/actions/runs/6228083319/job/16904088683

**Comment by antiagainst** [comment]
Filed https://github.com/openxla/iree/issues/15012 to track this. For now I'll disable the particular test to avoid marking all subsequent commits as failing CI.

**Pull request merged**